website/blog/2021-11-19-common-sub-expression-optimization.md

```diff
@@ -0,0 +1,80 @@
+
+![New Grouped PageProjection](/img/blog/2021-11-19-common-sub-expression-optimization/new-page-projection.png)
+
+The code that generates `PageProjection` is in `PageFunctionCompiler`. In this class, we first extract all CSEs, in the above example, there is only one CSE, which is `json_parse(features)`. We create a new assignment for the CSE, then rewrite each assignment using this newly introduced variable. So the assignments become:
+
+```
+cse := json_parse(features)
+expr := TRY_CAST(cse)
+expr_0 := TRY_CAST(cse)
+...
+expr_4 := TRY_CAST(cse)
+```
+
+Since all assignments depend on this CSE, we will wrap them in a single `PageProjection` class. We generate a function for cse that computes its value on first invocation and stores it, and on later invocation will just return the result directly.
+
+![PageFunctionCompiler](/img/blog/2021-11-19-common-sub-expression-optimization/page-function-compiler.png)
+
+In this way, the expensive `json_parse` will only need to be executed once.
+
+Let’s look at another example to get a better understanding of what counts as CSEs. In the following query:
+
+```
+SELECT
+    x + y + z,                        -- exp0
+    x + y * z,                        -- exp1
+    (x + y + z) * 2,                  -- exp2
+    cast(x + y + z as VARCHAR),       -- exp3
+    (x + y + z) * 2 * z               -- exp4
+FROM 
+    (VALUES (1, 2, 4), (3, 5, 7), (2, 4, 5)) t(x, y, z);
+```
+
+There are 5 assignments in the projection, with 2 CSEs: `x + y + z` and `(x + y + z) * 2` (`x + y` is also a CSE but all expressions using this also uses `x + y + z` so we do not generate a separate function for it). We can rewrite this projection into:
+
+```
+cse1 := x + y + z
+cse2 := cse1 * 2
+expr0 := cse1
+expr1 := x + y * z
+expr2 := cse2
+expr3 := CAST(cse1 as VARCHAR)
+expr4 := cse2 * z
+```
+
+From the rewritten assignments, we can see that expr1 does not share any CSEs with other expressions, and the rest do. We will generate 2 `PageProjection` classes, one for expr1, another one for the rest of the assignments.
+
+### Performance benefit
+
+Microbenchmark (`CommonSubExpessionBenchmark`) has shown that even for simple expressions like `x + y`, there could be a 10% performance improvement. For queries using complex expressions like `JSON_PARSE`, we’ve seen 3x performance improvements in production.
+
+![Benchmark](/img/blog/2021-11-19-common-sub-expression-optimization/benchmark.png)
+
+## How to use it?
+
+This optimization was first introduced in Presto release 0.235. There were subsequent improvements and bug fixes in the releases after that and was considered stable since release 0.245. The optimization is on by default. You can turn it off by setting session property `SET SESSION optimize_common_sub_expressions = false` to see the performance difference.
```

Comment:
```suggestion
This optimization was first introduced in Presto release 0.235. There were subsequent improvements and bug fixes in the releases after that, but it is considered stable since release 0.245. The optimization is on by default. You can turn it off by setting session property `SET SESSION optimize_common_sub_expressions = false` to see the performance difference.
```